File: backend/src/main/java/com/moeware/ims/entity/customer/Employee.java
package com.moeware.ims.entity.customer;

import jakarta.persistence.*;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import com.moeware.ims.entity.User;
import com.moeware.ims.entity.customer.Department;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "employees", indexes = {
        @Index(name = "idx_employees_code", columnList = "employee_code", unique = true),
        @Index(name = "idx_employees_email", columnList = "email", unique = true),
        @Index(name = "idx_employees_department", columnList = "department_id"),
        @Index(name = "idx_employees_manager", columnList = "manager_id"),
        @Index(name = "idx_employees_user", columnList = "user_id"),
        @Index(name = "idx_employees_name", columnList = "last_name, first_name"),
        @Index(name = "idx_employees_active", columnList = "id")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Employee {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank(message = "Employee code is required")
    @Size(max = 50, message = "Employee code must not exceed 50 characters")
    @Column(name = "employee_code", nullable = false, unique = true, length = 50)
    private String employeeCode;

    @NotBlank(message = "First name is required")
    @Size(max = 100, message = "First name must not exceed 100 characters")
    @Column(name = "first_name", nullable = false, length = 100)
    private String firstName;

    @NotBlank(message = "Last name is required")
    @Size(max = 100, message = "Last name must not exceed 100 characters")
    @Column(name = "last_name", nullable = false, length = 100)
    private String lastName;

    @NotBlank(message = "Email is required")
    @Email(message = "Invalid email format")
    @Size(max = 255, message = "Email must not exceed 255 characters")
    @Column(nullable = false, unique = true, length = 255)
    private String email;

    @Size(max = 50, message = "Phone must not exceed 50 characters")
    @Column(length = 50)
    private String phone;

    @Size(max = 100, message = "Job title must not exceed 100 characters")
    @Column(name = "job_title", length = 100)
    private String jobTitle;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "department_id")
    private Department department;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "manager_id")
    private Employee manager;

    @OneToMany(mappedBy = "manager", cascade = CascadeType.ALL)
    @Builder.Default
    private Set<Employee> subordinates = new HashSet<>();

    @NotNull(message = "Hire date is required")
    @Column(name = "hire_date", nullable = false)
    private LocalDate hireDate;

    @Column(name = "termination_date")
    private LocalDate terminationDate;

    @Column(precision = 12, scale = 2)
    private BigDecimal salary;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @Column(name = "is_active", nullable = false)
    @Builder.Default
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    // Helper methods
    public String getFullName() {
        return firstName + " " + lastName;
    }

    public boolean hasSystemAccess() {
        return user != null;
    }

    public boolean isManager() {
        return subordinates != null && !subordinates.isEmpty();
    }

    public int getSubordinateCount() {
        return subordinates != null ? subordinates.size() : 0;
    }

    public boolean isTerminated() {
        return terminationDate != null && !terminationDate.isAfter(LocalDate.now());
    }
}

File: frontend/src/types/employee.ts
// ========================================
// EMPLOYEE TYPES
// ========================================

// Lighter department reference (as returned by API)
export interface EmployeeDepartment {
  id: number
  name: string
  code: string
}

// Lighter warehouse reference (as returned by API)
export interface EmployeeWarehouse {
  id: number
  name: string
  code: string
}

// Lighter employee reference for manager (as returned by API)
export interface EmployeeManager {
  id: number
  firstName: string
  lastName: string
  email?: string
  employeeCode?: string
}

// System user reference
export interface EmployeeUser {
  id: number
  username: string
  role?: string
}

export interface Employee {
  id: number
  employeeCode: string
  firstName: string
  lastName: string
  email?: string
  phone?: string
  address?: string
  city?: string
  state?: string
  country?: string
  postalCode?: string
  department?: EmployeeDepartment
  warehouse?: EmployeeWarehouse
  manager?: EmployeeManager | null
  hireDate: string
  terminationDate?: string
  jobTitle?: string
  salary?: number
  isActive: boolean
  hasSystemAccess?: boolean
  user?: EmployeeUser
  createdAt: string
  updatedAt?: string
  version?: number
}

export interface EmployeeFormData {
  employeeCode: string
  firstName: string
  lastName: string
  email?: string
  phone?: string
  address?: string
  city?: string
  state?: string
  country?: string
  postalCode?: string
  departmentId?: number
  warehouseId?: number
  managerId?: number
  hireDate: string
  jobTitle?: string
  salary?: number
  isActive?: boolean
}
